### PR TITLE
[App] Enable to register data connections

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR refactors connect to add sub ressources:

```python
lightning connect data

lightning connect app

lightning connect X
```

NB: Need to merge some platforms stuffs for the tests to pass.

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda

### CODE REVIEW

**Review by akihironitta — APPROVED** [review]

File: src/lightning/app/cli/connect/data.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    name: str,
+    region: str,
+    source: str,
+    destination: str = "",
+    project_name: str = "",
+) -> None:
+    """Create a new data connection."""
+
+    if sys.platform == "win32":
+        print("`ls` isn't supported on windows. Open an issue on Github.")
```
Comment (ethanwharris) [inline src/lightning/app/cli/connect/data.py:47]:
```suggestion
        print("Data connection isn't supported on windows. Open an issue on Github.")
```

File: src/lightning/app/cli/connect/data.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+            if project.name == project_name:
+                project_id = project.project_id
+
+        if project_id is None:
+            project_id = _get_project(client).project_id
+
+        if not source.startswith("s3://"):
+            _error_and_exit(
+                "Only public s3 folder are supported for now. Please, open a Github issue with your use case."
```
Comment (ethanwharris) [inline src/lightning/app/cli/connect/data.py:69]:
```suggestion
                "Only public S3 folders are supported for now. Please open a Github issue with your use case."
```

File: src/lightning/app/cli/lightning_cli.py [hunk 3]
```diff
@@ -121,8 +122,21 @@ def show() -> None:
     pass
 
 
-_main.command()(connect)
-_main.command()(disconnect)
+@_main.group()
+def connect() -> None:
+    """Show resource connections."""
```
Comment (ethanwharris) [inline src/lightning/app/cli/lightning_cli.py:127]: `Show resource connections` doesn't sound right. Maybe like `Connect apps and data.`

File: src/lightning/app/cli/lightning_cli.py [hunk 4]
```diff
@@ -121,8 +122,21 @@ def show() -> None:
...
+    pass
+
+
+@_main.group()
+def disconnect() -> None:
+    """Show resource connections."""
```
Comment (ethanwharris) [inline src/lightning/app/cli/lightning_cli.py:133]: Same comment as above

**Review by ethanwharris — APPROVED** [review]
It feels a bit weird to me that `lightning connect data` and `lightning connect app` both exist in the same group as they are about two different kinds of connection. So `lightning connect app` doesn't feel right, it would be more like `lightning connect to app` - maybe `lightning connect` should be renamed entirely to `lightning attach` or something?

File: src/lightning/app/cli/connect/data.py (hunk 1 above)
Comment (tchaton) [reply src/lightning/app/cli/connect/data.py:47]: Good catch !

File: src/lightning/app/cli/connect/data.py (hunk 2 above)
Comment (tchaton) [reply src/lightning/app/cli/connect/data.py:69]: Done !

File: src/lightning/app/cli/lightning_cli.py (hunk 3 above)
Comment (tchaton) [reply src/lightning/app/cli/lightning_cli.py:127]: Done !

File: src/lightning/app/cli/lightning_cli.py (hunk 4 above)
Comment (tchaton) [reply src/lightning/app/cli/lightning_cli.py:133]: Done !

**Comment by tchaton** [comment]
> It feels a bit weird to me that `lightning connect data` and `lightning connect app` both exist in the same group as they are about two different kinds of connection. So `lightning connect app` doesn't feel right, it would be more like `lightning connect to app` - maybe `lightning connect` should be renamed entirely to `lightning attach` or something?

I agree. Let's re-align the CLI experience soon before Trainer 2.0. So we can improve on the wordings and consistency.

**Review by Borda — APPROVED** [review]

File: tests/tests_app/cli/test_connect_data.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+from unittest.mock import MagicMock
+
+import pytest
+from lightning_cloud.openapi import ProjectIdDataConnectionsBody, V1ListMembershipsResponse, V1Membership
+
+from lightning.app.cli.connect import data
+
+
+@pytest.mark.skipif(sys.platform == "win32", reason="not supported on windows yet")
+def test_connect_data(monkeypatch):
```
Comment (awaelchli) [inline tests/tests_app/cli/test_connect_data.py:11]: This test doesn't cover several code paths in the connect_data command.

File: src/lightning/app/cli/connect/data.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+        client = LightningClient()
+        projects = client.projects_service_list_memberships()
+
+        project_id = None
+
+        for project in projects.memberships:
+
+            if project.name == project_name:
+                project_id = project.project_id
```
Comment (awaelchli) [inline src/lightning/app/cli/connect/data.py:61]: If it is sufficient to find the first occurrence, consider adding a break here. I think probably the project name is unique for a user?

**Review by awaelchli — APPROVED** [review]
I agree with @ethanwharris

File: tests/tests_app/cli/test_connect_data.py (hunk 5 above)
Comment (tchaton) [reply tests/tests_app/cli/test_connect_data.py:11]: Done !

File: src/lightning/app/cli/connect/data.py (hunk 6 above)
Comment (tchaton) [reply src/lightning/app/cli/connect/data.py:61]: Done !

**Pull request merged**